drivers/serial/uart_stm32.c

```diff
@@ -2318,20 +2316,28 @@ static int uart_stm32_pm_action(const struct device *dev,
 		}
 
 		/* Enable clock */
-		err = clock_control_on(data->clock,
-					(clock_control_subsys_t)&config->pclken[0]);
+		err = clock_control_on(data->clock, (clock_control_subsys_t)&config->pclken[0]);
 		if (err < 0) {
 			LOG_ERR("Could not enable (LP)UART clock");
 			return err;
 		}
 
-		if ((IS_ENABLED(CONFIG_PM_S2RAM)) &&
-			(!LL_USART_IsEnabled(config->usart))) {
+		if (!LL_USART_IsEnabled(config->usart)) {
 			/* When exiting low power mode, check whether UART is enabled.
-			 * If not, it means we are exiting Suspend to RAM mode (STM32
-			 * Standby), and the driver needs to be reinitialized.
+			 * If not, it means that the peripheral configuration has been erased
+			 * by the low power mode.
+			 * Either do a full re-initialization after a Suspend to RAM mode (STM32
+			 * Standby), or simply configure the registers after a STOP2 mode.
```

Comment:
I would prefer a more explicit documentation. Looking at STM32WL5 RefMan (+ some unrelated rewording), I propose:
```suggestion
			 * If not, it means the peripheral has been powered down
			 * by the low-power mode. If suspend-to-RAM is enabled,
			 * assume the entire SoC has been powered down and do a
			 * full re-initialization. Otherwise, assume that the
			 * low-power mode shut down power to the UART but not
			 * critical peripherals (CPU, GPIO, RCC), which means
			 * we only have to reconfigure this UART instance.
			 *
			 * STOP2 on STM32WLE5 is an example of such low-power mode.
```